{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "407e7c3f",
   "metadata": {},
   "source": [
    "# COVID-19 TF-IDF and BM25 Based Information Retrieval System"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5c6a9e3d",
   "metadata": {},
   "source": [
    "Dalam notebook ini, kita membangun sistem pengambilan informasi sederhana menggunakan TF-IDF untuk menghasilkan array indeks terbalik, dan kemudian menggunakan BM25 untuk melakukan pencarian dan pengindeksan artikel.\n",
    "\n",
    "Langkah-langkah:\n",
    "1. Ekstraksi: Mengambil data dari dataset.\n",
    "2. Organisasi: Membuat dictionary untuk menyimpan informasi relevan dan menghasilkan indeks terbalik berdasarkan kata kunci dari setiap dokumen.\n",
    "3. Pengambilan: Menjalankan algoritma BM25 untuk menghasilkan daftar hasil yang telah diurutkan."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b591321d",
   "metadata": {},
   "source": [
    "## Langkah-Langkah Utama\n",
    "1. Ekstraksi Data\n",
    "Mengunduh file JSON dari koleksi dataset COVID-19, mengekstraknya, dan menyimpannya dalam folder data.\n",
    "\n",
    "2. Pra-pemrosesan Teks\n",
    "Melakukan:\n",
    "- Penghapusan stopword, tanda baca, angka, dan karakter khusus.\n",
    "- Konversi teks menjadi huruf kecil.\n",
    "- Stemming dan lemmatization.\n",
    "3. Ekstraksi Informasi\n",
    "Mengambil paper_id, judul, dan abstrak dari setiap artikel, lalu menyimpannya dalam DataFrame.\n",
    "\n",
    "4. Organisasi Data\n",
    "- Menghitung skor TF-IDF untuk setiap kata dalam abstrak.\n",
    "- Membuat indeks terbalik untuk mempercepat pencarian.\n",
    "5. Pengambilan Informasi\n",
    "- Menggunakan BM25 untuk menghitung relevansi artikel terhadap query.\n",
    "- Mengurutkan artikel berdasarkan skor relevansi."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "deb75039",
   "metadata": {},
   "source": [
    "## Contoh Query"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2c73ad9c",
   "metadata": {},
   "outputs": [],
   "source": [
    "q = ['asal-usul coronavirus', 'respons coronavirus terhadap perubahan cuaca', 'imunitas coronavirus']"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f7724308",
   "metadata": {},
   "source": [
    "Hasilnya adalah daftar artikel yang paling relevan untuk setiap query, diurutkan berdasarkan skor BM25."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "08402c84",
   "metadata": {},
   "source": [
    "## Library yang Digunakan  \n",
    "- Ekstraksi Data: os, urllib.request, tarfile, json, pandas, numpy, tqdm\n",
    "- Pra-pemrosesan Teks: re, nltk (stopwords, stemming, lemmatization)\n",
    "- Ekstraksi Kata Kunci: sklearn.feature_extraction.text, scipy.sparse\n",
    "- BM25: rank_bm25"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "d93e98f2",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[nltk_data] Downloading package stopwords to C:\\Users\\Tuhu\n",
      "[nltk_data]     Pangestu\\AppData\\Roaming\\nltk_data...\n",
      "[nltk_data]   Package stopwords is already up-to-date!\n",
      "[nltk_data] Downloading package wordnet to C:\\Users\\Tuhu\n",
      "[nltk_data]     Pangestu\\AppData\\Roaming\\nltk_data...\n",
      "[nltk_data]   Package wordnet is already up-to-date!\n"
     ]
    }
   ],
   "source": [
    "#libraries for getting data and extracting\n",
    "import os\n",
    "import urllib.request\n",
    "import tarfile\n",
    "import json\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from tqdm import tqdm\n",
    "\n",
    "\n",
    "#libraries for text preprocessing\n",
    "import re\n",
    "import nltk\n",
    "nltk.download('stopwords')\n",
    "from nltk.corpus import stopwords\n",
    "from nltk.stem.porter import PorterStemmer\n",
    "nltk.download('wordnet') \n",
    "from nltk.stem.wordnet import WordNetLemmatizer\n",
    "\n",
    "#libraries for keyword extraction with tf-idf\n",
    "from sklearn.feature_extraction.text import CountVectorizer\n",
    "from sklearn.feature_extraction.text import TfidfTransformer\n",
    "from scipy.sparse import coo_matrix\n",
    "\n",
    "#libraries for reading and writing files\n",
    "import pickle\n",
    "\n",
    "#libraries for BM25\n",
    "# !pip install rank_bm25\n",
    "from rank_bm25 import BM25Okapi"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "51d7707c",
   "metadata": {},
   "source": [
    "### Ekstraksi Data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4ebee4b8",
   "metadata": {},
   "source": [
    "Mengambil data\n",
    "\n",
    "Mengunduh file-file JSON dari koleksi dan menempatkannya ke dalam satu folder bernama data. Di dalam folder data, dibuat subfolder berdasarkan asal masing-masing artikel. Kita akan menggunakan file-file JSON yang disimpan dalam folder data ini pada langkah ekstraksi untuk membuat dataset kita."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "36f6ab13",
   "metadata": {},
   "outputs": [],
   "source": [
    "def getData():\n",
    "    urls = ['https://ai2-semanticscholar-cord-19.s3-us-west-2.amazonaws.com/2020-03-27/comm_use_subset.tar.gz', 'https://ai2-semanticscholar-cord-19.s3-us-west-2.amazonaws.com/2020-03-27/noncomm_use_subset.tar.gz', 'https://ai2-semanticscholar-cord-19.s3-us-west-2.amazonaws.com/2020-03-27/custom_license.tar.gz', 'https://ai2-semanticscholar-cord-19.s3-us-west-2.amazonaws.com/2020-03-27/biorxiv_medrxiv.tar.gz']\n",
    "\n",
    "    # Create data directory\n",
    "    try:\n",
    "        os.mkdir('./data')\n",
    "        print('Directory created')\n",
    "    except FileExistsError:\n",
    "        print('Directory already exists')\n",
    "\n",
    "    #Download all files\n",
    "    for i in range(len(urls)):\n",
    "        urllib.request.urlretrieve(urls[i], './data/file'+str(i)+'.tar.gz')\n",
    "        print('Downloaded file '+str(i+1)+'/'+str(len(urls)))\n",
    "        tar = tarfile.open('./data/file'+str(i)+'.tar.gz')\n",
    "        tar.extractall('./data')\n",
    "        tar.close()\n",
    "        print('Extracted file '+str(i+1)+'/'+str(len(urls)))\n",
    "        os.remove('./data/file'+str(i)+'.tar.gz')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "69d25733",
   "metadata": {},
   "source": [
    "### Pra-pemrosesan\n",
    "Input: Teks  \n",
    "Output: Teks yang telah dipra-proses – menghapus stopword, tanda baca, mengubah semua huruf menjadi huruf kecil, menghapus angka/karakter khusus, melakukan stemming, dan lemmatization.  \n",
    "\n",
    "Langkah ini diperlukan untuk tahap ekstraksi karena kita akan melakukan pra-pemrosesan terhadap semua teks yang diekstrak (judul dan abstrak)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "043307d5",
   "metadata": {},
   "outputs": [],
   "source": [
    "def preprocess(text):\n",
    "    #define stopwords\n",
    "    stop_words = list(stopwords.words('english'))\n",
    "    #Remove punctuations\n",
    "    text = re.sub('[^a-zA-Z]', ' ', text)\n",
    "    #Convert to lowercase\n",
    "    text = text.lower()\n",
    "    #remove tags\n",
    "    text=re.sub(\"&lt;/?.*?&gt;\",\" &lt;&gt; \",text)\n",
    "    # remove special characters and digits\n",
    "    text=re.sub(\"(\\\\d|\\\\W)+\",\" \",text)\n",
    "    ##Convert to list from string\n",
    "    text = text.split()\n",
    "    ##Stemming\n",
    "    ps=PorterStemmer()\n",
    "    text = [ps.stem(word) for word in text if not word in stop_words]\n",
    "    #Lemmatisation\n",
    "    lem = WordNetLemmatizer()\n",
    "    text = [lem.lemmatize(word) for word in text if not word in  stop_words] \n",
    "    text = \" \".join(text) \n",
    "    \n",
    "    return text"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "70b6e2e5",
   "metadata": {},
   "source": [
    "### Ekstraksi\n",
    "\n",
    "Input: Tidak ada  \n",
    "Output: Sebuah dataframe yang berisi semua informasi yang telah diekstrak  \n",
    "\n",
    "Untuk setiap artikel, kita mengumpulkan:\n",
    "- paper_id\n",
    "- Judul\n",
    "- Abstrak\n",
    "\n",
    "Kita menyimpan nilai-nilai ini ke dalam sebuah dataframe menggunakan pustaka pandas, lalu menyimpannya ke dalam file pickle."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "44d61389",
   "metadata": {},
   "outputs": [],
   "source": [
    "def extract():\n",
    "    #create our collection locally in the data folder\n",
    "    \n",
    "    #creating our initial datastructure\n",
    "    x = {'paper_id':[], 'title':[], 'abstract': []}\n",
    "    \n",
    "    #Iterate through all files in the data directory\n",
    "    for subdir, dirs, files in os.walk('./data'):\n",
    "        for file in tqdm(files):\n",
    "            with open(os.path.join(subdir, file)) as f:\n",
    "                data = json.load(f)\n",
    "                \n",
    "               #Append paper ID to list\n",
    "                x['paper_id'].append(data['paper_id'])\n",
    "               #Append article title to list & preprocess the text\n",
    "                x['title'].append((data['metadata']['title']))\n",
    "                \n",
    "                #Append abstract text content values only to abstract list & preprocess the text\n",
    "                abstract = \"\"\n",
    "                for paragraph in data['abstract']:\n",
    "                    abstract += paragraph['text']\n",
    "                    abstract += '\\n'\n",
    "                #if json file no abstract in file, set the body text as the abstract (happens rarely, but often enough that this edge case matters)\n",
    "                if abstract == \"\": \n",
    "                    for paragraph in data['body_text']:\n",
    "                        abstract += paragraph['text']\n",
    "                        abstract += '\\n'\n",
    "                x['abstract'].append(preprocess(abstract))\n",
    "                \n",
    "    #Create Pandas dataframe & write to pickle file\n",
    "    df = pd.DataFrame.from_dict(x, orient='index')\n",
    "    df = df.transpose()\n",
    "    pickle.dump( df, open( \"full_data_processed_FINAL.p\", \"wb\" ) )\n",
    "    return df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0b95ec22",
   "metadata": {},
   "source": [
    "**Organize**\n",
    "\n",
    "Urutkan Matriks Koordinat  \n",
    "Input: representasi koordinat tf_idf  \n",
    "Output: item tf_idf yang diurutkan dalam urutan menurun — sehingga item dengan skor tertinggi berada di atas!  \n",
    "\n",
    "Fungsi untuk mengurutkan tf_idf dalam urutan menurun."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "09e7535e",
   "metadata": {},
   "outputs": [],
   "source": [
    "def sort_coo(coo_matrix):\n",
    "    tuples = zip(coo_matrix.col, coo_matrix.data)\n",
    "    return sorted(tuples, key=lambda x: (x[1], x[0]), reverse=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8575837c",
   "metadata": {},
   "source": [
    "Mengambil Top N Kata dengan Skor TF‑IDF Tertinggi\n",
    "Input:\n",
    "1. feature_names = daftar kosakata\n",
    "2. sorted_items = vektor TF‑IDF yang sudah diurutkan menurun\n",
    "3. topN = jumlah kata kunci yang ingin Anda ekstrak dari teks\n",
    "\n",
    "Dictionary berisi topN kata dengan skor TF‑IDF tertinggi dalam teks (sebagai key) beserta skor TF‑IDF masing‑masing (sebagai value)  \n",
    "Fungsi ini mengembalikan nama kata kunci dan skor TF‑IDF mereka untuk topN item teratas."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "995bb75d",
   "metadata": {},
   "outputs": [],
   "source": [
    "def extract_topn_from_vector(feature_names, sorted_items, topN):\n",
    "    #use only topn items from vector\n",
    "    sorted_items = sorted_items[:topN]\n",
    " \n",
    "    score_vals = []\n",
    "    feature_vals = []\n",
    "    # word index and corresponding tf-idf score\n",
    "    for idx, score in sorted_items:\n",
    "        #keep track of feature name and its corresponding score\n",
    "        score_vals.append(round(score, 3))\n",
    "        feature_vals.append(feature_names[idx])\n",
    " \n",
    "    #create a tuples of feature,score\n",
    "    results= {}\n",
    "    for idx in range(len(feature_vals)):\n",
    "        results[feature_vals[idx]]=score_vals[idx]\n",
    "    return results"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e63d707d",
   "metadata": {},
   "source": [
    "Mengambil Kata Kunci dari Abstrak  \n",
    "\n",
    "Input:\n",
    "1. entry = baris dalam DataFrame artikel, yang mewakili satu artikel\n",
    "2. cv = objek CountVectorizer dari sklearn.feature_extraction.text\n",
    "3. X = matriks hasil CountVectorizer yang sudah di-fit pada korpus\n",
    "4. tfidf_transformer = objek yang menyimpan data TF‑IDF — sudah di-fit pada korpus\n",
    "5. feature_names = daftar kosakata\n",
    "6. topN = jumlah kata kunci yang ingin diambil dari abstrak\n",
    "\n",
    "Output:\n",
    "TopN kata kunci teratas dari abstrak  \n",
    "  \n",
    "Fungsi ini mengekstrak dan mengembalikan topN kata kunci dengan skor TF‑IDF tertinggi dari abstrak."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "baca826c",
   "metadata": {},
   "outputs": [],
   "source": [
    "def getAbstractKeywords(entry, cv, X, tfidf_transformer, feature_names, topN):\n",
    "    abstract = entry['abstract']\n",
    "    \n",
    "    #first check that abstract is full\n",
    "    if type(abstract) == float:\n",
    "        return []\n",
    " \n",
    "    #generate tf-idf for the given document\n",
    "    tf_idf_vector=tfidf_transformer.transform(cv.transform([abstract])) \n",
    "    #sort the tf-idf vectors by descending order of scores\n",
    "    sorted_items=sort_coo(tf_idf_vector.tocoo())\n",
    "    #extract only the topN # items\n",
    "    keywords_dict=extract_topn_from_vector(feature_names,sorted_items,topN)\n",
    "    #just want words themselves, so only need keys of the dictionary\n",
    "    keywords = list(keywords_dict.keys()) \n",
    "     \n",
    "    return keywords"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "820c0c93",
   "metadata": {},
   "source": [
    "Mengambil Kata Kunci dari Judul  \n",
    "**Input:**  \n",
    "entry = baris dalam DataFrame artikel, yang mewakili satu artikel  \n",
    "**Output:**    \n",
    "Daftar semua kata yang terdapat dalam judul  \n",
    "  \n",
    "Diandaikan bahwa jika sebuah kata terdapat dalam judul artikel, maka kata tersebut penting untuk artikel dan diperlakukan sebagai kata kunci. Oleh karena itu, metode ini hanya mengekstrak semua kata dari judul yang telah diproses sebelumnya."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "2bee2241",
   "metadata": {},
   "outputs": [],
   "source": [
    "def getTitleKeywords(entry):\n",
    "    title = entry['title']  \n",
    "    title = preprocess(title)\n",
    "    #first check that the title of that entry is full\n",
    "    if type(title) == float:\n",
    "        return []\n",
    "    \n",
    "    keywords_title = title.split(' ')\n",
    "    return keywords_title"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "46609d33",
   "metadata": {},
   "source": [
    "Mengambil keyword final  \n",
    "Input:  \n",
    "\n",
    "1. entry = baris dalam dataframe artikel, yang merepresentasikan satu artikel\n",
    "2. cv = CountVectorizer dari sklearn.feature_extraction.text\n",
    "3. X = vektor yang merepresentasikan hasil fit CountVectorizer terhadap korpus\n",
    "4. tfidf_transformer = objek yang menyimpan data tf-idf — juga sudah difit ke korpus\n",
    "5. feature_names = daftar kosakata\n",
    "6. topN = jumlah keyword yang ingin diekstrak dari abstrak\n",
    "\n",
    "Output:  \n",
    "Daftar semua keyword untuk sebuah artikel — diekstrak dari judul dan abstrak!\n",
    "\n",
    "Fungsi ini memanggil getTitleKeywords() dan getAbstractKeywords(), lalu menggabungkan kedua daftar tersebut menjadi satu daftar keyword final."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "baa82e03",
   "metadata": {},
   "outputs": [],
   "source": [
    "def getFinalKeywords(entry, cv, X, tfidf_trans, feature_names, topN):\n",
    "    #get keywords from abstract and title\n",
    "    fromAbstract = getAbstractKeywords(entry, cv, X, tfidf_trans, feature_names, topN)\n",
    "    fromTitle = getTitleKeywords(entry)\n",
    "    #concatenate two lists\n",
    "    finalKeywords = fromAbstract + fromTitle\n",
    "    #convert to set and then back to list to ensure there are no duplicates in list\n",
    "    final_no_duplicates = list(set(finalKeywords))\n",
    "    return final_no_duplicates"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c2b304ff",
   "metadata": {},
   "source": [
    "Mengambil korpus\n",
    "Input:  \n",
    "Sebuah dataframe yang berisi:  \n",
    "1. paper_id\n",
    "2. abstract\n",
    "3. article (untuk setiap artikel. Semua teks sudah diproses.)\n",
    "  \n",
    "Output:  \n",
    "Sebuah daftar (list) yang berisi seluruh abstrak dalam dataframe artikel.  \n",
    "  \n",
    "Langkah ini digunakan untuk membuat corpus, yang merupakan input penting dalam proses tf-idf kita."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "179d4423",
   "metadata": {},
   "outputs": [],
   "source": [
    "def getCorpus(articlesDf):\n",
    "    #creating a new dataframe, abstractDf, of just the abstracts, so that we don't modify the original dataframe, articlesDf\n",
    "    abstractDf = pd.DataFrame(columns = ['abstract'])\n",
    "    #filling abstractDf with the abstract column from articlesDf\n",
    "    abstractDf['abstract'] = articlesDf['abstract']\n",
    "    #converting column of dataframe to a list\n",
    "    corpus = abstractDf['abstract'].to_list()\n",
    "    return corpus\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ffe3a4d1",
   "metadata": {},
   "source": [
    "**Menambahkan keyword**  \n",
    "Input:  \n",
    "1. df = DataFrame yang berisi:  \n",
    "    - paper_id\n",
    "    - title\n",
    "    - abstract (untuk setiap artikel. Semua teks sudah diproses.)\n",
    "\n",
    "2. top N = jumlah keyword yang ingin diekstrak dari abstrak\n",
    "3. makeFile = nilai boolean, menentukan apakah fungsi ini akan membuat file pickle dari dataframe hasil\n",
    "4. fileName = nama file pickle yang diinginkan oleh pengguna\n",
    "  \n",
    "Output:  \n",
    "Sebuah DataFrame pandas yang berisi:  \n",
    "1. paper_id\n",
    "2. title\n",
    "3. abstract\n",
    "4. keyword yang dikaitkan dengan setiap artikel"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "9f9c8e4d",
   "metadata": {},
   "outputs": [],
   "source": [
    "def addKeywords(df, topN, makeFile, fileName):\n",
    "    #defining stopwords\n",
    "    stop_words = list(stopwords.words(\"english\"))\n",
    "\n",
    "    #creating following variables that are needed for keyword extract from abstract, using tf-idf methodology,\n",
    "    #all input in getFinalKewords method\n",
    "    corpus = getCorpus(df)\n",
    "    cv=CountVectorizer(max_df=0.8,stop_words=stop_words, max_features=1000, ngram_range=(1,1))    \n",
    "    X=cv.fit_transform(corpus)\n",
    "    tfidf_transformer=TfidfTransformer(smooth_idf=True,use_idf=True)\n",
    "    tfidf_transformer.fit(X)\n",
    "    feature_names=cv.get_feature_names_out()\n",
    "    \n",
    "    #adding keywords article to dataframe\n",
    "    df = df.reindex(columns = ['paper_id', 'title', 'abstract','keywords'])                \n",
    "    #getting keywords for each entry in article dataframe -- using apply to be more efficient\n",
    "    df['keywords'] = df.apply(lambda row: getFinalKeywords(row, cv, X, tfidf_transformer, feature_names, topN), axis=1)\n",
    "\n",
    "    #make pickle file depending on user input\n",
    "    if makeFile == True:\n",
    "        pickle.dump( df, open( fileName, \"wb\" ) )\n",
    "    return df  "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1123f547",
   "metadata": {},
   "source": [
    "Membuat inverted indices  \n",
    "Input:  \n",
    "Sebuah DataFrame pandas yang berisi:  \n",
    "1. paper_id\n",
    "2. title\n",
    "3. abstract\n",
    "4. keywords yang dikaitkan dengan setiap artikel\n",
    "  \n",
    "Output:  \n",
    "Sebuah dictionary dari inverted indices  \n",
    "- key = kata yang merupakan keyword\n",
    "- value = semua paper_id dari artikel yang memiliki kata tersebut sebagai keyword\n",
    "  \n",
    "Fungsi ini membuat dictionary inverted index, yang nantinya akan digunakan saat menentukan subset artikel untuk menjalankan algoritma pencarian atau peringkat. Output-nya penting berupa dictionary karena memiliki waktu pencarian yang konstan (constant look-up time)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "6d6972a3",
   "metadata": {},
   "outputs": [],
   "source": [
    "def createInvertedIndices(df):\n",
    "    numEntries = df.shape[0]\n",
    "    invertInd = {}\n",
    "    \n",
    "    for i in range (numEntries):\n",
    "        entry = df.iloc[i]\n",
    "        paper_id = entry['paper_id']    \n",
    "        keywords = entry['keywords']\n",
    "        for k in keywords:\n",
    "            if k not in invertInd:\n",
    "                invertInd[k] = []\n",
    "                invertInd[k].append(paper_id)\n",
    "            else:\n",
    "                invertInd[k].append(paper_id)\n",
    "    return invertInd"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d64530b6",
   "metadata": {},
   "source": [
    "organize step"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "cb37556c",
   "metadata": {},
   "outputs": [],
   "source": [
    "def organize():\n",
    "    df_without_keywords = pickle.load(open(\"full_data_processed_FINAL.p\", \"rb\"))\n",
    "    df_with_keywords = addKeywords(df_without_keywords, 10, False, \"full_data_withKeywords_FINAL.p\")\n",
    "    invertedIndices = createInvertedIndices(df_with_keywords)\n",
    "    pickle.dump( invertedIndices, open( \"invertedIndices_FINAL.p\", \"wb\" ) )"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3e24284a",
   "metadata": {},
   "source": [
    "**Data setelah diproses**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "6fd75a98",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>paper_id</th>\n",
       "      <th>title</th>\n",
       "      <th>abstract</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0015023cc06b5362d332b3baf348d11567ca2fbb</td>\n",
       "      <td>The RNA pseudoknots in foot-and-mouth disease ...</td>\n",
       "      <td>word count text word count author funder right...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>00340eea543336d54adda18236424de6a5e91c9d</td>\n",
       "      <td>Analysis Title: Regaining perspective on SARS-...</td>\n",
       "      <td>past three month new coronaviru sar cov epidem...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>004f0f8bb66cf446678dc13cf2701feec4f36d76</td>\n",
       "      <td>Healthcare-resource-adjusted vulnerabilities t...</td>\n",
       "      <td>ncov epidem spread across china countri februa...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>00d16927588fb04d4be0e6b269fc02f0d3c2aa7b</td>\n",
       "      <td>Real-time, MinION-based, amplicon sequencing f...</td>\n",
       "      <td>infecti bronchiti ib caus signific econom loss...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0139ea4ca580af99b602c6435368e7fdbefacb03</td>\n",
       "      <td>A Combined Evidence Approach to Prioritize Nip...</td>\n",
       "      <td>nipah viru niv came limelight recent due outbr...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33370</th>\n",
       "      <td>ff5a79ed22ea416e6d89caad1cf0d83dbc741a4b</td>\n",
       "      <td>Understanding Human Coronavirus HCoV-NL63</td>\n",
       "      <td>even though coronaviru infect human normal ass...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33371</th>\n",
       "      <td>ff6d57f2aad99be129432058665b361dc18747e8</td>\n",
       "      <td>Brief Definitive Report MACROPHAGES GENETICALL...</td>\n",
       "      <td>extens evid cultur macrophag genet resist susc...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33372</th>\n",
       "      <td>ff83907653a4c4500e8c509ca28169e924742b40</td>\n",
       "      <td>Identification of a Subdomain of CENPB That Is...</td>\n",
       "      <td>combin vivo vitro approach investig function c...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33373</th>\n",
       "      <td>ffe718db1820f27bf274e3fc519ab78e450de288</td>\n",
       "      <td>Replication enhancer elements within the open ...</td>\n",
       "      <td>provid experiment evid replic enhanc element r...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33374</th>\n",
       "      <td>ffed5d2a31a0c1a0db11905fe378e7735b6d70ca</td>\n",
       "      <td>Supplemental material for the paper \"Evidence ...</td>\n",
       "      <td>israel correspond author tt tamirtul post tau ...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>33375 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                       paper_id  \\\n",
       "0      0015023cc06b5362d332b3baf348d11567ca2fbb   \n",
       "1      00340eea543336d54adda18236424de6a5e91c9d   \n",
       "2      004f0f8bb66cf446678dc13cf2701feec4f36d76   \n",
       "3      00d16927588fb04d4be0e6b269fc02f0d3c2aa7b   \n",
       "4      0139ea4ca580af99b602c6435368e7fdbefacb03   \n",
       "...                                         ...   \n",
       "33370  ff5a79ed22ea416e6d89caad1cf0d83dbc741a4b   \n",
       "33371  ff6d57f2aad99be129432058665b361dc18747e8   \n",
       "33372  ff83907653a4c4500e8c509ca28169e924742b40   \n",
       "33373  ffe718db1820f27bf274e3fc519ab78e450de288   \n",
       "33374  ffed5d2a31a0c1a0db11905fe378e7735b6d70ca   \n",
       "\n",
       "                                                   title  \\\n",
       "0      The RNA pseudoknots in foot-and-mouth disease ...   \n",
       "1      Analysis Title: Regaining perspective on SARS-...   \n",
       "2      Healthcare-resource-adjusted vulnerabilities t...   \n",
       "3      Real-time, MinION-based, amplicon sequencing f...   \n",
       "4      A Combined Evidence Approach to Prioritize Nip...   \n",
       "...                                                  ...   \n",
       "33370          Understanding Human Coronavirus HCoV-NL63   \n",
       "33371  Brief Definitive Report MACROPHAGES GENETICALL...   \n",
       "33372  Identification of a Subdomain of CENPB That Is...   \n",
       "33373  Replication enhancer elements within the open ...   \n",
       "33374  Supplemental material for the paper \"Evidence ...   \n",
       "\n",
       "                                                abstract  \n",
       "0      word count text word count author funder right...  \n",
       "1      past three month new coronaviru sar cov epidem...  \n",
       "2      ncov epidem spread across china countri februa...  \n",
       "3      infecti bronchiti ib caus signific econom loss...  \n",
       "4      nipah viru niv came limelight recent due outbr...  \n",
       "...                                                  ...  \n",
       "33370  even though coronaviru infect human normal ass...  \n",
       "33371  extens evid cultur macrophag genet resist susc...  \n",
       "33372  combin vivo vitro approach investig function c...  \n",
       "33373  provid experiment evid replic enhanc element r...  \n",
       "33374  israel correspond author tt tamirtul post tau ...  \n",
       "\n",
       "[33375 rows x 3 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>paper_id</th>\n",
       "      <th>title</th>\n",
       "      <th>abstract</th>\n",
       "      <th>keywords</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0015023cc06b5362d332b3baf348d11567ca2fbb</td>\n",
       "      <td>The RNA pseudoknots in foot-and-mouth disease ...</td>\n",
       "      <td>word count text word count author funder right...</td>\n",
       "      <td>[genom, count, product, replic, nucleotid, len...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>00340eea543336d54adda18236424de6a5e91c9d</td>\n",
       "      <td>Analysis Title: Regaining perspective on SARS-...</td>\n",
       "      <td>past three month new coronaviru sar cov epidem...</td>\n",
       "      <td>[trace, base, perspect, molecular, light, anal...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>004f0f8bb66cf446678dc13cf2701feec4f36d76</td>\n",
       "      <td>Healthcare-resource-adjusted vulnerabilities t...</td>\n",
       "      <td>ncov epidem spread across china countri februa...</td>\n",
       "      <td>[ncov, healthcar, local, health, public, numbe...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>00d16927588fb04d4be0e6b269fc02f0d3c2aa7b</td>\n",
       "      <td>Real-time, MinION-based, amplicon sequencing f...</td>\n",
       "      <td>infecti bronchiti ib caus signific econom loss...</td>\n",
       "      <td>[time, serotyp, upper, use, lineag, viru, rt, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0139ea4ca580af99b602c6435368e7fdbefacb03</td>\n",
       "      <td>A Combined Evidence Approach to Prioritize Nip...</td>\n",
       "      <td>nipah viru niv came limelight recent due outbr...</td>\n",
       "      <td>[pain, base, diver, inhibitor, drug, combin, r...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33370</th>\n",
       "      <td>ff5a79ed22ea416e6d89caad1cf0d83dbc741a4b</td>\n",
       "      <td>Understanding Human Coronavirus HCoV-NL63</td>\n",
       "      <td>even though coronaviru infect human normal ass...</td>\n",
       "      <td>[child, mainli, infect, hcov, infant, respirat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33371</th>\n",
       "      <td>ff6d57f2aad99be129432058665b361dc18747e8</td>\n",
       "      <td>Brief Definitive Report MACROPHAGES GENETICALL...</td>\n",
       "      <td>extens evid cultur macrophag genet resist susc...</td>\n",
       "      <td>[macrophag, vitro, viru, genet, report, labora...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33372</th>\n",
       "      <td>ff83907653a4c4500e8c509ca28169e924742b40</td>\n",
       "      <td>Identification of a Subdomain of CENPB That Is...</td>\n",
       "      <td>combin vivo vitro approach investig function c...</td>\n",
       "      <td>[subdomain, suffici, local, dna, protein, repe...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33373</th>\n",
       "      <td>ffe718db1820f27bf274e3fc519ab78e450de288</td>\n",
       "      <td>Replication enhancer elements within the open ...</td>\n",
       "      <td>provid experiment evid replic enhanc element r...</td>\n",
       "      <td>[mutant, replic, read, frame, enhanc, infect, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33374</th>\n",
       "      <td>ffed5d2a31a0c1a0db11905fe378e7735b6d70ca</td>\n",
       "      <td>Supplemental material for the paper \"Evidence ...</td>\n",
       "      <td>israel correspond author tt tamirtul post tau ...</td>\n",
       "      <td>[lambda, count, multipl, region, paramet, effi...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>33375 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                       paper_id  \\\n",
       "0      0015023cc06b5362d332b3baf348d11567ca2fbb   \n",
       "1      00340eea543336d54adda18236424de6a5e91c9d   \n",
       "2      004f0f8bb66cf446678dc13cf2701feec4f36d76   \n",
       "3      00d16927588fb04d4be0e6b269fc02f0d3c2aa7b   \n",
       "4      0139ea4ca580af99b602c6435368e7fdbefacb03   \n",
       "...                                         ...   \n",
       "33370  ff5a79ed22ea416e6d89caad1cf0d83dbc741a4b   \n",
       "33371  ff6d57f2aad99be129432058665b361dc18747e8   \n",
       "33372  ff83907653a4c4500e8c509ca28169e924742b40   \n",
       "33373  ffe718db1820f27bf274e3fc519ab78e450de288   \n",
       "33374  ffed5d2a31a0c1a0db11905fe378e7735b6d70ca   \n",
       "\n",
       "                                                   title  \\\n",
       "0      The RNA pseudoknots in foot-and-mouth disease ...   \n",
       "1      Analysis Title: Regaining perspective on SARS-...   \n",
       "2      Healthcare-resource-adjusted vulnerabilities t...   \n",
       "3      Real-time, MinION-based, amplicon sequencing f...   \n",
       "4      A Combined Evidence Approach to Prioritize Nip...   \n",
       "...                                                  ...   \n",
       "33370          Understanding Human Coronavirus HCoV-NL63   \n",
       "33371  Brief Definitive Report MACROPHAGES GENETICALL...   \n",
       "33372  Identification of a Subdomain of CENPB That Is...   \n",
       "33373  Replication enhancer elements within the open ...   \n",
       "33374  Supplemental material for the paper \"Evidence ...   \n",
       "\n",
       "                                                abstract  \\\n",
       "0      word count text word count author funder right...   \n",
       "1      past three month new coronaviru sar cov epidem...   \n",
       "2      ncov epidem spread across china countri februa...   \n",
       "3      infecti bronchiti ib caus signific econom loss...   \n",
       "4      nipah viru niv came limelight recent due outbr...   \n",
       "...                                                  ...   \n",
       "33370  even though coronaviru infect human normal ass...   \n",
       "33371  extens evid cultur macrophag genet resist susc...   \n",
       "33372  combin vivo vitro approach investig function c...   \n",
       "33373  provid experiment evid replic enhanc element r...   \n",
       "33374  israel correspond author tt tamirtul post tau ...   \n",
       "\n",
       "                                                keywords  \n",
       "0      [genom, count, product, replic, nucleotid, len...  \n",
       "1      [trace, base, perspect, molecular, light, anal...  \n",
       "2      [ncov, healthcar, local, health, public, numbe...  \n",
       "3      [time, serotyp, upper, use, lineag, viru, rt, ...  \n",
       "4      [pain, base, diver, inhibitor, drug, combin, r...  \n",
       "...                                                  ...  \n",
       "33370  [child, mainli, infect, hcov, infant, respirat...  \n",
       "33371  [macrophag, vitro, viru, genet, report, labora...  \n",
       "33372  [subdomain, suffici, local, dna, protein, repe...  \n",
       "33373  [mutant, replic, read, frame, enhanc, infect, ...  \n",
       "33374  [lambda, count, multipl, region, paramet, effi...  \n",
       "\n",
       "[33375 rows x 4 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# getData()\n",
    "# extract()\n",
    "\n",
    "df_without_keywords = pickle.load(open(\"full_data_processed_FINAL.p\", \"rb\"))\n",
    "df_with_keywords = addKeywords(df_without_keywords, 10, False, \"full_data_withKeywords_FINAL.p\")\n",
    "display(df_without_keywords)\n",
    "display(df_with_keywords)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0a014f57",
   "metadata": {},
   "source": [
    "RETRIEVE"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5988cd88",
   "metadata": {},
   "source": [
    "mendapatkan subset artikel kita\n",
    "Input: query\n",
    "Output: daftar artikel potensial yang mungkin relevan, karena memiliki beberapa istilah query sebagai kata kunci mereka\n",
    "\n",
    "Langkah ini dilakukan agar kita tidak perlu menjalankan algoritma peringkat, BM25, pada semua ~30.000 artikel. Tujuannya adalah untuk mengidentifikasi subset ini sehingga proses peringkat (dan pengambilan) menjadi lebih efisien!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "466856b5",
   "metadata": {},
   "outputs": [],
   "source": [
    "def getPotentialArticleSubset(query):\n",
    "    #load in inverted indices\n",
    "    invertedIndices = pickle.load(open(\"invertedIndices_FINAL.p\", \"rb\"))\n",
    "    \n",
    "    #preprocess query and split into individual terms\n",
    "    query = preprocess(query)\n",
    "    queryTerms = query.split(' ')\n",
    "    \n",
    "    potentialArticles = []\n",
    "    #concatenate list of potential articles by looping through potential articles for each word in query\n",
    "    for word in queryTerms:\n",
    "        if word in invertedIndices: #so if someone types in nonsensical query term that's not in invertedIndices, still won't break!\n",
    "            someArticles = invertedIndices[word]\n",
    "            potentialArticles = potentialArticles + someArticles\n",
    "            \n",
    "    #convert to set then back to list so there are no repeat articles\n",
    "    potentialArticles = list(set(potentialArticles))\n",
    "    return potentialArticles"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "194319fd",
   "metadata": {},
   "source": [
    "### BM25\n",
    "\n",
    "Input: daftar artikel, dictionary yang berisi semua dokumen, bobot untuk judul, bobot untuk abstrak, dan query\n",
    "Output: daftar artikel yang telah diurutkan\n",
    "\n",
    "Ini adalah metode utama untuk pengambilan informasi yang mengimplementasikan algoritma Okapi BM25."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "b43803f5",
   "metadata": {},
   "outputs": [],
   "source": [
    "def bm25(articles, df_dic, title_w, abstract_w, query):\n",
    "    corpus_title = []\n",
    "    corpus_abstract = []\n",
    "    \n",
    "    for article in articles:\n",
    "        arr = df_dic.get(article)\n",
    "        #title\n",
    "        if type(arr[0]) != float:\n",
    "            preprocessedTitle = preprocess(arr[0])\n",
    "            corpus_title.append(preprocessedTitle)\n",
    "        else:\n",
    "            corpus_title.append(\" \")\n",
    "        \n",
    "        #abstract\n",
    "        if type(arr[1]) != float:\n",
    "            preprocessedAbst = preprocess(arr[1])\n",
    "            corpus_abstract.append(preprocessedAbst)\n",
    "        else:\n",
    "            corpus_abstract.append(\" \")\n",
    "            \n",
    "    query = preprocess(query)\n",
    "    \n",
    "    tokenized_query = query.split(\" \")\n",
    "    \n",
    "    tokenized_corpus_title = [doc.split(\" \") for doc in corpus_title]\n",
    "    tokenized_corpus_abstract = [doc.split(\" \") for doc in corpus_abstract]\n",
    "    \n",
    "    #running bm25 on titles\n",
    "    bm25_title = BM25Okapi(tokenized_corpus_title)\n",
    "    doc_scores_titles = bm25_title.get_scores(tokenized_query)\n",
    "    #weighting array\n",
    "    doc_scores_titles = np.array(doc_scores_titles)\n",
    "    doc_scores_titles = doc_scores_titles**title_w\n",
    "    \n",
    "    #running bm25 on abstracts\n",
    "    bm25_abstract = BM25Okapi(tokenized_corpus_abstract)\n",
    "    doc_scores_abstracts = bm25_abstract.get_scores(tokenized_query)\n",
    "    #weighting\n",
    "    doc_scores_abstracts = np.array(doc_scores_abstracts)\n",
    "    doc_scores_abstracts = doc_scores_abstracts ** abstract_w\n",
    "    \n",
    "    #summing up the two different scores\n",
    "    doc_scores = np.add(doc_scores_abstracts,doc_scores_titles)\n",
    "    \n",
    "    #creating a dictionary with the scores\n",
    "    score_dict = dict(zip(articles, doc_scores))\n",
    "    \n",
    "    #creating list of ranked documents high to low\n",
    "    doc_ranking = sorted(score_dict, key=score_dict.get, reverse = True)\n",
    "    \n",
    "    #get top 100\n",
    "    doc_ranking = doc_ranking[0:100]\n",
    "    \n",
    "    for i in range(len(doc_ranking)):\n",
    "        dic_entry = df_dic.get(doc_ranking[i])\n",
    "        doc_ranking[i] = dic_entry[0]\n",
    "    \n",
    "    return doc_ranking"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4f8bf5da",
   "metadata": {},
   "source": [
    "retrieval step"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "ceebe6f5",
   "metadata": {},
   "outputs": [],
   "source": [
    "def retrieve(queries):\n",
    "    #performing information retrieval\n",
    "    df_without_keywords = pickle.load(open(\"full_data_processed_FINAL.p\", \"rb\"))\n",
    "    df_dic = df_without_keywords.set_index('paper_id').T.to_dict('list')\n",
    "    results = []\n",
    "    for q in queries:\n",
    "        articles = getPotentialArticleSubset(q)\n",
    "        result = bm25(articles,df_dic,1,2,q)\n",
    "        results.append(result)\n",
    "\n",
    "    #Output results\n",
    "    for query in range(len(results)):\n",
    "        for rank in range(len(results[query])):\n",
    "            print(str(query+1)+'\\t'+str(rank+1)+'\\t'+str(results[query][rank]))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aa180aa2",
   "metadata": {},
   "source": [
    "Menggabungkan semuanya :  \n",
    "Ini akan memberikan kita 100 artikel teratas untuk setiap query yang diurutkan berdasarkan peringkat BM25."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "33ec0d0f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Directory already exists\n",
      "Downloaded file 1/4\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\Tuhu Pangestu\\AppData\\Local\\Temp\\ipykernel_11496\\2087884789.py:16: DeprecationWarning: Python 3.14 will, by default, filter extracted tar archives and reject files or modify their metadata. Use the filter argument to control this behavior.\n",
      "  tar.extractall('./data')\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Extracted file 1/4\n",
      "Downloaded file 2/4\n",
      "Extracted file 2/4\n",
      "Downloaded file 3/4\n",
      "Extracted file 3/4\n",
      "Downloaded file 4/4\n",
      "Extracted file 4/4\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "0it [00:00, ?it/s]\n",
      "100%|██████████| 1053/1053 [00:43<00:00, 24.40it/s]\n",
      "100%|██████████| 9315/9315 [06:19<00:00, 24.52it/s]\n",
      "100%|██████████| 20657/20657 [17:22<00:00, 19.81it/s] \n",
      "100%|██████████| 2350/2350 [02:22<00:00, 16.47it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1\t1\tInsights into the cross-species evolution of 2019 novel coronavirus\n",
      "1\t2\tBat origin of human coronaviruses\n",
      "1\t3\t\n",
      "1\t4\tOrigin, Evolution, and Genotyping of Emergent Porcine Epidemic Diarrhea Virus Strains in the United States\n",
      "1\t5\tThe battle against SARS and MERS coronaviruses: Reservoirs and Animal Models\n",
      "1\t6\tNewly emerged porcine enteric alphacoronavirus in southern China: Identification, origin and evolutionary history analysis\n",
      "1\t7\tMolecular Evolution of Human Coronavirus Genomes\n",
      "1\t8\tMOTS CLÉS\n",
      "1\t9\tBat origin of a new human coronavirus: there and back again\n",
      "1\t10\tA review of studies on animal reservoirs of the SARS coronavirus\n",
      "1\t11\tSARS-Coronavirus ancestor's foot-prints in South-East Asian bat colonies and the refuge theory\n",
      "1\t12\tTesting the hypothesis of a recombinant origin of the SARS-associated coronavirus\n",
      "1\t13\tScienceDirect Focus on Middle East respiratory syndrome coronavirus (MERS-CoV)\n",
      "1\t14\t\n",
      "1\t15\tEvidence of recombination in coronaviruses implicating pangolin origins of nCoV- 2019\n",
      "1\t16\tCoronaviruses: emerging and re-emerging pathogens in humans and animals\n",
      "1\t17\tEmergence of SARS-CoV-2 through Recombination and Strong Purifying Selection Short Title: Recombination and origin of SARS-CoV-2 One Sentence Summary: Extensive Recombination and Strong Purifying Selection among coronaviruses from different hosts facilitate the emergence of SARS-CoV-2\n",
      "1\t18\tRecombination and Coronavirus Defective Interfering RNAs\n",
      "1\t19\tCOVID-19 infection: origin, transmission, and characteristics of human coronaviruses\n",
      "1\t20\tA Bat-Derived Putative Cross-Family Recombinant Coronavirus with a Reovirus Gene a11111 OPEN ACCESS\n",
      "1\t21\tLes infections à coronavirus humains\n",
      "1\t22\tPCR sequencing of the spike genes of geographically and chronologically distinct human coronaviruses 229E\n",
      "1\t23\tGenomic characterisation and epidemiology of 2019 novel coronavirus: implications for virus origins and receptor binding\n",
      "1\t24\tDirect diagnosis of human respiratory coronaviruses 229E and OC43 by the polymerase chain reaction\n",
      "1\t25\tSpike Protein Fusion Peptide and Feline Coronavirus Virulence\n",
      "1\t26\tGenomic Analysis and Surveillance of the Coronavirus Dominant in Ducks in China\n",
      "1\t27\tEvidence from the evolutionary analysis of nucleotide sequences for a recombinant history of SARS-CoV\n",
      "1\t28\tRecombination in Avian Gamma-Coronavirus Infectious Bronchitis Virus\n",
      "1\t29\tBiomedical Science Review SARS Virus: The Beginning of the Unraveling of a New Coronavirus\n",
      "1\t30\tComplete genome sequence of bat coronavirus HKU2 from Chinese horseshoe bats revealed a much smaller spike gene with a different evolutionary lineage from the rest of the genome\n",
      "1\t31\tComplete Genome Sequence of a Novel Swine Acute Diarrhea Syndrome Coronavirus, CH/FJWT/2018, Isolated in Fujian, China, in 2018\n",
      "1\t32\tIdentification and Characterization of a Novel Alpaca Respiratory Coronavirus Most Closely Related to the Human Coronavirus 229E\n",
      "1\t33\tREPLICATION OF A BOVINE CORONAVIRUS IN ORGAN CULTURES OF FOETAL TRACHEA\n",
      "1\t34\t\n",
      "1\t35\tRecombinant Canine Coronaviruses in Dogs, Europe\n",
      "1\t36\tHuman Coronavirus EMC Is Not the Same as Severe Acute Respiratory Syndrome Coronavirus\n",
      "1\t37\tDiscovery of Bat Coronaviruses through Surveillance and Probe Capture-Based Next-Generation Sequencing\n",
      "1\t38\tSelection of and recombination between minor variants lead to the adaptation of an avian coronavirus to primate cells\n",
      "1\t39\tA second, non-canonical RNA-dependent RNA polymerase in SARS Coronavirus\n",
      "1\t40\tMiddle East Respiratory Syndrome Coronavirus (MERS-CoV) origin and animal reservoir\n",
      "1\t41\tCryo-electron microscopy structure of the SADS-CoV spike glycoprotein 1 provides insights into an evolution of unique coronavirus spike proteins 2 3\n",
      "1\t42\tJournal Pre-proof The SARS-CoV-2 outbreak: what we know The SARS-CoV-2 Outbreak: What We Know\n",
      "1\t43\tGenomic characterization of the 2019 novel human-pathogenic coronavirus isolated from a patient with atypical pneumonia after visiting Wuhan\n",
      "1\t44\tOutbreak 2019-nCoV (Wuhan virus), a novel Coronavirus: human-to-human transmission, travel-related cases, and vaccine readiness\n",
      "1\t45\tTropical Medicine and Infectious Disease Potential Intermediate Hosts for Coronavirus Transmission: No Evidence of Clade 2c Coronaviruses in Domestic Livestock from Ghana\n",
      "1\t46\t\n",
      "1\t47\tSARS associated coronavirus has a recombinant polymerase and coronaviruses have a history of host-shifting\n",
      "1\t48\tDetection of alpha-and betacoronaviruses in rodents from Yunnan, China\n",
      "1\t49\tThe aetiology, origins, and diagnosis of severe acute respiratory syndrome\n",
      "1\t50\tDetection and full genome characterization of two beta CoV viruses related to Middle East respiratory syndrome from bats in Italy\n",
      "1\t51\tGenetic diversity of a canine coronavirus detected in pups with diarrhoea in Italy\n",
      "1\t52\tDiscovery, diversity and evolution of novel coronaviruses sampled from rodents in China\n",
      "1\t53\tThe nucleocapsid protein of the SARS coronavirus is capable of self-association through a C-terminal 209 amino acid interaction domain q\n",
      "1\t54\t\n",
      "1\t55\tGeographical structure of bat SARS-related coronaviruses\n",
      "1\t56\tBiopolymeric nano/microspheres for selective and reversible adsorption of coronaviruses\n",
      "1\t57\tUnderstanding the accessory viral proteins unique to the severe acute respiratory syndrome (SARS) coronavirus\n",
      "1\t58\tHuman-animal interactions and bat coronavirus spillover potential among rural residents in Southern China-NC-ND license (http://creativecommons.org/licenses/by-nc-nd/4.0/)\n",
      "1\t59\tStructural Insights into the Interaction of Coronavirus Papain-Like Proteases and Interferon-Stimulated Gene Product 15 from Different Species\n",
      "1\t60\tDetection of potentially novel paramyxovirus and coronavirus viral RNA in bats and rats in the Mekong Delta region of southern Viet Nam\n",
      "1\t61\tOpinion Novel coronavirus: From discovery to clinical diagnostics\n",
      "1\t62\tMolecular and phylogenetic analysis of bovine coronavirus based on the spike glycoprotein gene\n",
      "1\t63\tCanine respiratory coronavirus employs caveolin-1-mediated pathway for internalization to HRT-18G cells\n",
      "1\t64\tDiscovery of a novel canine respiratory coronavirus support genetic recombination among betacoronavirus1\n",
      "1\t65\t\n",
      "1\t66\tSARS-CoV-2 is sensitive to type I interferon pretreatment\n",
      "1\t67\tAdaptation of human enteric coronavirus to growth in cell lines\n",
      "1\t68\tMolecular characterization of a canine respiratory coronavirus strain detected in Italy ଝ\n",
      "1\t69\tEpidemiology, Genetic Recombination, and Pathogenesis of Coronaviruses\n",
      "1\t70\tCryo-EM structures of HKU2 and SADS-CoV spike 1 glycoproteins and insights into coronavirus evolution 2 3 4\n",
      "1\t71\tNo credible evidence supporting claims of the laboratory engineering of SARS-CoV-2\n",
      "1\t72\tDevelopment of a Whole-Virus ELISA for Serological Evaluation of Domestic Livestock as Possible Hosts of Human Coronavirus NL63\n",
      "1\t73\tX-ray Structure of Main Protease of the Novel Coronavirus SARS-CoV-2 Enables Design of a-Ketoamide Inhibitors\n",
      "1\t74\tCross-sectional survey of selected enteric viruses in Polish turkey flocks between 2008 and 2011\n",
      "1\t75\tEnzyme-linked Immunosorbent Assay for the Detection of Porcine Epidemic Diarrhea Coronavirus Antibodies in Swine Sera\n",
      "1\t76\tArticle Bat Origins of MERS-CoV Supported by Bat Coronavirus HKU4 Usage of Human Receptor CD26\n",
      "1\t77\tFunctional analysis of the stem loop S3 and S4 structures in the coronavirus 3′UTR\n",
      "1\t78\tNew Alphacoronavirus in Mystacina tuberculata Bats, New Zealand\n",
      "1\t79\tIdentification of New Respiratory Viruses in the New Millennium\n",
      "1\t80\tGenomic characterization and infectivity of a novel SARS-like coronavirus in Chinese bats\n",
      "1\t81\tMemory T cell responses targeting the SARS coronavirus persist up to 11 years post-infection\n",
      "1\t82\t\n",
      "1\t83\tFurther Evidence for Bats as the Evolutionary Source of Middle East Respiratory Syndrome Coronavirus\n",
      "1\t84\tPutative Receptor Binding Domain of Bat-Derived Coronavirus HKU9 Spike Protein: Evolution of Betacoronavirus Receptor Binding Motifs\n",
      "1\t85\tPorcine epidemic diarrhea: A retrospect from Europe and matters of debate\n",
      "1\t86\tCell adhesion as a novel approach to determining the cellular binding motif on the severe acute respiratory syndrome coronavirus spike protein\n",
      "1\t87\tOrigin and Possible Genetic Recombination of the Middle East Respiratory Syndrome Coronavirus from the First Imported Case in China: Phylogenetics and Coalescence Analysis\n",
      "1\t88\tClinical Medicine Editorial Initial Cluster of Novel Coronavirus (2019-nCoV) Infections in Wuhan, China Is Consistent with Substantial Human-to-Human Transmission\n",
      "1\t89\tShared Common Ancestry of Rodent Alphacoronaviruses Sampled Globally\n",
      "1\t90\tNEW RESPIRATORY VIRUS IN PEDIATRICS\n",
      "1\t91\tAnimal origins of SARS coronavirus: possible links with the international trade in small carnivores\n",
      "1\t92\tMERS Coronavirus: An Emerging Zoonotic Virus\n",
      "1\t93\tEmergence of the Middle East Respiratory Syndrome Coronavirus\n",
      "1\t94\tReplication of MERS and SARS coronaviruses in bat cells offers insights to their ancestral origins\n",
      "1\t95\tPerspectives on the epizootiology of feline enteric coronavirus and the pathogenesis of feline infectious peritonitis\n",
      "1\t96\t\n",
      "1\t97\tClinical Microbiology Middle Eastern Respiratory Syndrome Coronavirus (MERS-CoV)\n",
      "1\t98\tFull-genome evolutionary analysis of the novel corona virus (2019-nCoV) rejects the hypothesis of emergence as a result of a recent recombination event\n",
      "1\t99\tImmunogenicity of an adenoviral-based Middle East Respiratory Syndrome coronavirus vaccine in BALB/c mice\n",
      "1\t100\tNOTE Feline coronavirus isolates from a part of Brazil: insights into molecular epidemiology and phylogeny inferred from the 7b gene\n",
      "2\t1\tA case-crossover analysis of the impact of weather on primary cases of Middle East respiratory syndrome\n",
      "2\t2\tAdvice-giving in newspaper weather commentaries\n",
      "2\t3\tSevere weather warnings predict fracture epidemics\n",
      "2\t4\tSYSTEMATIC REVIEW Extreme water-related weather events and waterborne disease\n",
      "2\t5\tPathogen seasonality and links with weather in England and Wales: a big data time series analysis\n",
      "2\t6\tShort Term Effects of Weather on Hand, Foot and Mouth Disease\n",
      "2\t7\tInfluence of extreme weather and meteorological anomalies on outbreaks of influenza A (H1N1)\n",
      "2\t8\tJournal Pre-proof A conceptual model for the outbreak of Coronavirus disease 2019 (COVID-19) in Wuhan, China with individual reaction and governmental action\n",
      "2\t9\tClimate Change and Health in Canada\n",
      "2\t10\tWeather-Dependent Risk for Legionnaires' Disease, United States Legionella pneumophila [leʺjə-nelʹə nooʺmo-filʹə]\n",
      "2\t11\tImpacts of Social and Economic Factors on the Transmission of Coronavirus Disease 2019 (COVID-19) in China *\n",
      "2\t12\tComputational Statistics and Data Analysis Modeling respiratory illnesses with change point: A lesson from the SARS epidemic in Hong Kong Change point problem Multiple index model Non-parametric regression Public health Respiratory illness Severe acute respiratory syndrome (SARS)\n",
      "2\t13\tChallenges in developing methods for quantifying the effects of weather and climate on water-associated diseases: A systematic review\n",
      "2\t14\tChina's adaptation to climate & urban climatic changes: A critical review\n",
      "2\t15\tImpact of global warming on viral diseases: what is the evidence? This review comes from a themed issue on Pharmaceutical biotechnology Edited\n",
      "2\t16\t\n",
      "2\t17\tClimate change and infectious diseases: What can we expect?\n",
      "2\t18\tRAT CORONAVIRUS\n",
      "2\t19\tRespiratory syncytial virus bronchiolitis, weather conditions and air pollution in an Italian urban area: An observational study\n",
      "2\t20\t\n",
      "2\t21\tNum 1\n",
      "2\t22\tOne Health: Zoonoses in the Exotic Animal Practice KEYWORDS • Zoonoses • Exotic pets • Nontraditional pets • Wildlife • Disease prevention\n",
      "2\t23\tMapping Climate Change Vulnerabilities to Infectious Diseases in Europe\n",
      "2\t24\t\n",
      "2\t25\tBovine Coronavirus Associated Syndromes\n",
      "2\t26\t\n",
      "2\t27\tCoronaviridae\n",
      "2\t28\tMillennium Project The Millennium Project: the positive health implications of improved environmental sustainability\n",
      "2\t29\t\n",
      "2\t30\tTitle: Effects of weather-related social distancing on city-scale transmission of respiratory viruses (limit 96 characters)\n",
      "2\t31\tTransportation and\n",
      "2\t32\tEmerging and Re-emerging Pathogens and Diseases, and Health Consequences of a Changing Climate\n",
      "2\t33\tPublic Health Response Systems In-Action: Learning from Local Health Departments' Experiences with Acute and Emergency Incidents\n",
      "2\t34\tEnvironmental Health, Planetary Boundaries and Limits to Growth\n",
      "2\t35\tStabilized coronavirus spikes are resistant to conformational changes induced by receptor recognition or proteolysis OPEN\n",
      "2\t36\tPCR sequencing of the spike genes of geographically and chronologically distinct human coronaviruses 229E\n",
      "2\t37\tMechanisms of Coronavirus Cell Entry Mediated by the Viral Spike Protein\n",
      "2\t38\tCharacterization of amino acid substitutions in feline coronavirus 3C-like protease from a cat with feline infectious peritonitis treated with a protease inhibitor\n",
      "2\t39\tSpike Protein Fusion Peptide and Feline Coronavirus Virulence\n",
      "2\t40\tFatal Enteritis Associated with Coronavirus Infection in Cats\n",
      "2\t41\tCorrespondence Submissions should be made via our electronic submission system at http://ees.elsevier.com/ thelancet\n",
      "2\t42\tMolecular Modeling Evaluation of the Binding Effect of Ritonavir, Lopinavir and Darunavir to Severe Acute Respiratory Syndrome Coronavirus 2 Proteases\n",
      "2\t43\tStructure of mouse coronavirus spike protein complexed with receptor reveals mechanism for viral entry\n",
      "2\t44\t\n",
      "2\t45\t\n",
      "2\t46\tComparative sequence analysis of full-length genome of FIPV at different tissue passage levels\n",
      "2\t47\tA Massachusetts prototype like coronavirus isolated from wild peafowls is pathogenic to chickens\n",
      "2\t48\tEvidence for RNA editing in the transcriptome of 2019 Novel Coronavirus\n",
      "2\t49\tCanine Coronavirus Highly Pathogenic for Dogs\n",
      "2\t50\tCaution: The clinical characteristics of COVID-19 patients at admission are changing\n",
      "2\t51\tQuasispecies of bovine enteric and respiratory coronaviruses based on complete genome sequences and genetic changes after tissue culture adaptation\n",
      "2\t52\tAnalysis on the Clinical Characteristics of 36 Cases of Novel Coronavirus Pneumonia in Kunming\n",
      "2\t53\tEmergence of a group 3 coronavirus through recombination\n",
      "2\t54\tAltering SARS Coronavirus Frameshift Efficiency Affects Genomic and Subgenomic RNA Production\n",
      "2\t55\tSystemic fatal type II coronavirus infection in a dog: Pathological findings and immunohistochemistry\n",
      "2\t56\t\n",
      "2\t57\t\n",
      "2\t58\tPotent inhibition of feline coronaviruses with peptidyl compounds targeting coronavirus 3C-like protease\n",
      "2\t59\tPrediction of New Coronavirus Infection Based on a Modified SEIR Model\n",
      "2\t60\tTopographic Changes in SARS Coronavirus-infected Cells during Late Stages of Infection\n",
      "2\t61\tBat origin of a new human coronavirus: there and back again\n",
      "2\t62\tViral and Cellular mRNA Translation in Coronavirus-Infected Cells\n",
      "2\t63\tComparison of the S genes and the biological properties of respiratory and enteropathogenic bovine coronaviruses Brief Report\n",
      "2\t64\tReversal of the Progression of Fatal Coronavirus Infection in Cats by a Broad- Spectrum Coronavirus Protease Inhibitor\n",
      "2\t65\tRecombination in Avian Gamma-Coronavirus Infectious Bronchitis Virus\n",
      "2\t66\tStructural basis for human coronavirus attachment to sialic acid receptors HHS Public Access Author manuscript\n",
      "2\t67\tVirus Research Molecular characterization of the S proteins of two enterotropic murine coronavirus strains\n",
      "2\t68\tMurine encephalitis caused by HCoV-OC43, a human coronavirus with broad species specificity, is partly immune-mediated\n",
      "2\t69\tSevere acute respiratory syndrome-related coronavirus: The species and its viruses -a statement of the Coronavirus Study Group\n",
      "2\t70\tCharacterizations of Coronavirus cis-Acting RNA Elements and the Transcription Step Affecting Its Transcription Efficiency\n",
      "2\t71\tSARS associated coronavirus has a recombinant polymerase and coronaviruses have a history of host-shifting\n",
      "2\t72\tSelection of and recombination between minor variants lead to the adaptation of an avian coronavirus to primate cells\n",
      "2\t73\tDecline of transmissible gastroenteritis virus and its complex evolutionary relationship with porcine respiratory coronavirus in the United states\n",
      "2\t74\tViral and Cellular mRNA Translation in Coronavirus- Infected Cells\n",
      "2\t75\tInferring the hosts of coronavirus using dual statistical models based on nucleotide composition OPEN\n",
      "2\t76\tIntestinal replication of a porcine respiratory coronavirus closely related antigenically to the enteric transmissible gastroenteritis virus\n",
      "2\t77\tPotential for global spread of a novel coronavirus from China\n",
      "2\t78\tApplying lessons from SARS to a newly identifi ed coronavirus\n",
      "2\t79\tBMC Infectious Diseases Date of origin of the SARS coronavirus strains\n",
      "2\t80\tBovine coronaviruses associated with enteric and respiratory diseases in Canadian dairy cattle display different reactivities to anti-HE monoclonal antibodies and distinct amino acid changes in their HE, S and ns4.9 protein 7 0 2 ( 0 1 ) 0 0 2 4 3 -X\n",
      "2\t81\tConsensus statement The species Severe acute respiratory syndrome- related coronavirus: classifying 2019-nCoV and naming it SARS-CoV-2 Coronaviridae Study Group of the International Committee on Taxonomy of Viruses*\n",
      "2\t82\tMINIREVIEW Coronavirus Spike Proteins in Viral Entry and Pathogenesis\n",
      "2\t83\tTWIRLS, an automated topic-wise inference method based on massive literature, suggests a possible mechanism via ACE2 for the pathological changes in the human host after coronavirus infection\n",
      "2\t84\tGlobal Climate Changes and International Trade and Travel: Effects on Human Health Outcomes\n",
      "2\t85\tVirus-based vectors for gene expression in mammalian cells: Coronavirus\n",
      "2\t86\t\n",
      "2\t87\tTWIRLS, an automated topic-wise inference method based on massive literature, suggests a possible mechanism via ACE2 for the pathological changes in the human host after coronavirus infection\n",
      "2\t88\tCellular composition, coronavirus antigen expression and production of specific antibodies in lesions in feline infectious peritonitis\n",
      "2\t89\tProteomics analysis of differentially expressed proteins in chicken trachea and kidney after infection with the highly virulent and attenuated coronavirus infectious bronchitis virus in vivo\n",
      "2\t90\tSpringer Seminars in Immunopathology Immunopathological aspects of coronavirus infections\n",
      "2\t91\tThe human coronavirus HCoV-229E S-protein structure and receptor binding\n",
      "2\t92\tThe aetiology of SARS: Koch's postulates fulfilled\n",
      "2\t93\tSARS-coronavirus replicates in mononuclear cells of peripheral blood (PBMCs) from SARS patients\n",
      "2\t94\t\n",
      "2\t95\tCanine respiratory coronavirus employs caveolin-1-mediated pathway for internalization to HRT-18G cells\n",
      "2\t96\tMolecular and phylogenetic analysis of bovine coronavirus based on the spike glycoprotein gene\n",
      "2\t97\tDirectional and reoccurring sequence change in zoonotic RNA virus genomes visualized by time- series word count OPEN\n",
      "2\t98\tSerological Responses in Patients with Severe Acute Respiratory Syndrome Coronavirus Infection and Cross-Reactivity with Human Coronaviruses 229E, OC43, and NL63\n",
      "2\t99\tSTUDIES WITH AN UNCLASSIFIED VIRUS ISOLATED FROM DIARRHEIC CALVES\n",
      "2\t100\tVaccines for emerging infectious diseases: Lessons from MERS coronavirus and Zika virus\n",
      "3\t1\tSARS CORONAVIRUS AND INNATE IMMUNITY\n",
      "3\t2\tSARS coronavirus and innate immunity\n",
      "3\t3\tTitle: Candidate targets for immune responses to 2019-Novel Coronavirus (nCoV): sequence homology-and bioinformatic-based predictions\n",
      "3\t4\tCoronavirus membrane-associated papain-like proteases induce autophagy through interacting with Beclin1 to negatively regulate antiviral innate immunity\n",
      "3\t5\tToll-Like Receptor 3 Signaling via TRIF Contributes to a Protective Innate Immune Response to Severe Acute Respiratory Syndrome Coronavirus Infection\n",
      "3\t6\tIdentification of a Gamma Interferon-Activated Inhibitor of Translation-Like RNA Motif at the 3= End of the Transmissible Gastroenteritis Coronavirus Genome Modulating Innate Immune Response\n",
      "3\t7\tThe Immunity Induced by Recombinant Spike Proteins of SARS Coronavirus in Balb/c Mice\n",
      "3\t8\tThe Conserved Coronavirus Macrodomain Promotes Virulence and Suppresses the Innate Immune Response during Severe Acute Respiratory Syndrome Coronavirus Infection\n",
      "3\t9\tProjecting the transmission dynamics of SARS-CoV-2 through the post-pandemic period\n",
      "3\t10\tMolecular pathology of emerging coronavirus infections\n",
      "3\t11\tpathogens Epigenetic Landscape during Coronavirus Infection\n",
      "3\t12\tVulnerabilities in coronavirus glycan shields despite extensive glycosylation\n",
      "3\t13\tA Sequence Homology and Bioinformatic Approach Can Predict Candidate Targets for Immune Responses to SARS-CoV-2 Theory A Sequence Homology and Bioinformatic Approach Can Predict Candidate Targets for Immune Responses to SARS-CoV-2\n",
      "3\t14\tMedical Microbiology and Immunology Reactivity of Antibodies in Human Serum with Antigens of an Enteropathogenic Bovine Coronavirus\n",
      "3\t15\tDendritic Cell-Specific Antigen Delivery by Coronavirus Vaccine Vectors Induces Long-Lasting Protective Antiviral and Antitumor Immunity\n",
      "3\t16\tT-cell immunity of SARS-CoV: Implications for vaccine development against MERS-CoV\n",
      "3\t17\t\n",
      "3\t18\tPerspectives on the epizootiology of feline enteric coronavirus and the pathogenesis of feline infectious peritonitis\n",
      "3\t19\tCD8+ T Cells Responding to the Middle East Respiratory Syndrome Coronavirus Nucleocapsid Protein Delivered by Vaccinia Virus MVA in Mice\n",
      "3\t20\tHuman CD4 + memory T-lymphocyte responses to SARS coronavirus infection\n",
      "3\t21\tEarly endonuclease-mediated evasion of RNA sensing ensures efficient coronavirus replication\n",
      "3\t22\tNatural Infection with the Porcine Respiratory Coronavirus Induces Protective Lactogenic Immunity against Transmissible Gastroenteritis\n",
      "3\t23\tBiomedical Science Review SARS Virus: The Beginning of the Unraveling of a New Coronavirus\n",
      "3\t24\tMurine Coronavirus-Induced Encephalomyelitides in Rats: Analysis of Immunoglobulins and Virus-Specific Antibodies in Serum and Cerebrospinal Fluid\n",
      "3\t25\tReversal of the Progression of Fatal Coronavirus Infection in Cats by a Broad- Spectrum Coronavirus Protease Inhibitor\n",
      "3\t26\tEuropean Journal of Case Reports in Internal Medicine Severe Immune Thrombocytopenia Complicated by Intracerebral Haemorrhage Associated with Coronavirus Infection: A Case Report and Literature Review\n",
      "3\t27\tIdentification of murine CD8 T cell epitopes in codon-optimized SARS-associated coronavirus spike protein\n",
      "3\t28\tCryo-electron microscopy structure of the SADS-CoV spike glycoprotein 1 provides insights into an evolution of unique coronavirus spike proteins 2 3\n",
      "3\t29\tSelective Packaging in Murine Coronavirus Promotes Virulence by Limiting Type I Interferon Responses\n",
      "3\t30\tInterferon and cytokine responses to SARS-coronavirus infection\n",
      "3\t31\tImmune evasion of porcine enteric coronaviruses and viral modulation of antiviral innate signaling\n",
      "3\t32\tEnteric Viral Infections of Calves and Passive Immunity 1\n",
      "3\t33\tIntragastric administration of Lactobacillus casei expressing transmissible gastroentritis coronavirus spike glycoprotein induced specific antibody production\n",
      "3\t34\tAssembly and immunogenicity of coronavirus-like particles carrying infectious bronchitis virus M and S proteins\n",
      "3\t35\tAntagonism of dsRNA-Induced Innate Immune Pathways by NS4a and NS4b Accessory Proteins during MERS Coronavirus Infection\n",
      "3\t36\tA genome-wide association study identifies major loci affecting the immune response against infectious bronchitis virus in chicken\n",
      "3\t37\tDendritic Cell-Specific Delivery of Flt3L by Coronavirus Vectors Secures Induction of Therapeutic Antitumor Immunity\n",
      "3\t38\t\n",
      "3\t39\tProtection from Lethal Coronavirus Infection by Affinity-Purified Spike Glycoprotein of Murine Hepatitis Virus, Strain A59\n",
      "3\t40\tExpression profiles of immune mediators in feline Coronavirus-infected cells and clinical samples of feline Coronavirus-positive cats\n",
      "3\t41\tTransmissible Gastroenteritis Virus Papain-Like Protease 1 Antagonizes Production of Interferon-through Its Deubiquitinase Activity\n",
      "3\t42\tUniversity of North Carolina at Chapel Hill, Chapel Hill, NC, and accepted by Editorial Board Member Linda J\n",
      "3\t43\tAttempted immunisation of cats against feline infectlous peritonitis using canine coronavirus\n",
      "3\t44\tArchives of Virology Studies on the Relationship Between Coronaviruses from the Intestinal and Respiratory Tracts oi Calves\n",
      "3\t45\tTransmissible gastroenteritis virus (TGEV)-based vectors with engineered murine tropism express the rotavirus VP7 protein and immunize mice against rotavirus\n",
      "3\t46\tCorrespondence\n",
      "3\t47\tFerret Coronavirus- Associated Diseases\n",
      "3\t48\tRAT CORONAVIRUS\n",
      "3\t49\t\n",
      "3\t50\tSuperior immune responses induced by intranasal immunization with recombinant adenovirus-based vaccine expressing full- length Spike protein of Middle East respiratory syndrome coronavirus\n",
      "3\t51\tThe immune vulnerability landscape of the 2019 Novel Coronavirus, SARS- CoV-2\n",
      "3\t52\tStructural insights into the Middle East respiratory syndrome coronavirus 4a protein and its dsRNA binding mechanism OPEN\n",
      "3\t53\tCoronavirus-induced ER stress response and its involvement in regulation of coronavirus-host interactions\n",
      "3\t54\tMolecular characterization of a virulent canine coronavirus BGF strain ଝ\n",
      "3\t55\tShort Communication\n",
      "3\t56\tIntestinal protection against challenge with transmissible gastroenteritis virus of pigs immune after infection with the porcine respiratory coronavirus\n",
      "3\t57\tUnlocking bat immunology: establishment of Pteropus alecto bone marrow-derived dendritic cells and macrophages OPEN\n",
      "3\t58\tLY6E impairs coronavirus fusion and confers immune control of viral disease\n",
      "3\t59\tComplement Activation Contributes to Severe Acute Respiratory Syndrome Coronavirus Pathogenesis\n",
      "3\t60\tImmunogenicity of an adenoviral-based Middle East Respiratory Syndrome coronavirus vaccine in BALB/c mice\n",
      "3\t61\tAntibody-dependent infection of human macrophages by severe acute respiratory syndrome coronavirus\n",
      "3\t62\tExcretion and persistence of bovine coronavirus in neonatal calves Brief Report\n",
      "3\t63\tOrchitis in a Cat Associated with Coronavirus Infection\n",
      "3\t64\tA Novel Bacterium-Like Particle Vaccine Displaying the MERS-CoV Receptor-Binding Domain Induces Specific Mucosal and Systemic Immune Responses in Mice\n",
      "3\t65\t\n",
      "3\t66\tMiddle East Respiratory Syndrome Coronavirus NS4b Protein Inhibits Host RNase L Activation\n",
      "3\t67\tMolecular mechanisms of coronavirus RNA capping and methylation\n",
      "3\t68\tAntibodies against trimeric S glycoprotein protect hamsters against SARS-CoV challenge despite their capacity to mediate Fc␥RII-dependent entry into B cells in vitro\n",
      "3\t69\t\n",
      "3\t70\tAugmentation of immune responses to SARS coronavirus by a combination of DNA and whole killed virus vaccines\n",
      "3\t71\tCoronavirus immunogens\n",
      "3\t72\tDetection of an Antigenic Group 2 Coronavirus in an Adult Alpaca with Enteritis ᰔ Downloaded from\n",
      "3\t73\tDetection of an Antigenic Group 2 Coronavirus in an Adult Alpaca with Enteritis ᰔ Downloaded from\n",
      "3\t74\tSARS coronavirus pathogenesis: host innate immune responses and viral antagonism of interferon\n",
      "3\t75\tProtection from SARS coronavirus conferred by live measles vaccine expressing the spike glycoprotein\n",
      "3\t76\t\n",
      "3\t77\tIntranasal immunization with inactivated SARS-CoV (SARS-associated coronavirus) induced local and serum antibodies in mice\n",
      "3\t78\tInduction of lactogenic immunity to transmissible gastroenteritis virus of swine using an attenuated coronavirus mutant able to survive in the physicochemical environment of the digestive tract\n",
      "3\t79\t\n",
      "3\t80\tDNA vaccine encoding Middle East respiratory syndrome coronavirus S1 protein induces protective immune responses in mice\n",
      "3\t81\tGlucose-6-Phosphate Dehydrogenase Enhances Antiviral Response through Downregulation of NADPH Sensor HSCARG and Upregulation of NF-κB Signaling\n",
      "3\t82\tMolecular targets for the rational design of drugs to inhibit SARS coronavirus\n",
      "3\t83\tEnteric Viral Infections of Pigs and Strategies for Induction of Mucosal Immunity\n",
      "3\t84\tDNA Vaccine Encoding Middle East Respiratory Syndrome Coronavirus S1 Protein Induces Protective Immune Responses in Mice HHS Public Access\n",
      "3\t85\tImmunogenicity, safety, and protective efficacy of an inactivated SARS-associated coronavirus vaccine in rhesus monkeys\n",
      "3\t86\tSwine acute diarrhea syndrome coronavirus (SADS-CoV) antagonizes interferon-β production via blocking IPS-1 and RIG-I\n",
      "3\t87\tMolecular mechanisms of primary and secondary mucosal immunity using avian infectious bronchitis virus as a model system\n",
      "3\t88\tSublingual immunization with recombinant adenovirus encoding SARS-CoV spike protein induces systemic and mucosal immunity without redirection of the virus to the brain\n",
      "3\t89\tInduction of Specific Immune Responses by Severe Acute Respiratory Syndrome Coronavirus Spike DNA Vaccine with or without Interleukin-2 Immunization Using Different Vaccination Routes in Mice ᰔ\n",
      "3\t90\tMERS-CoV 4b protein interferes with the NF- κB-dependent innate immune response during infection Author summary\n",
      "3\t91\tSARS-CoV Regulates Immune Function-Related Gene Expression in Human Monocytic Cells\n",
      "3\t92\t\n",
      "3\t93\tEnhanced protection in mice induced by immunization with inactivated whole viruses compare to spike protein of middle east respiratory syndrome coronavirus\n",
      "3\t94\tHuman β-defensin 2 plays a regulatory role in innate antiviral immunity and is capable of potentiating the induction of antigen- specific immunity\n",
      "3\t95\tProtective Role of Toll-like Receptor 3-Induced Type I Interferon in Murine Coronavirus Infection of Macrophages\n",
      "3\t96\tTreatment of cats with feline infectious peritonitis\n",
      "3\t97\t\n",
      "3\t98\tDevelopment of a Western Blot Assay for Detection of Antibodies against Coronavirus Causing Severe Acute Respiratory Syndrome\n",
      "3\t99\tAdenoviral expression of a truncated S1 subunit of SARS-CoV spike protein results in specific humoral immune responses against SARS-CoV in rats\n",
      "3\t100\tVirus-based vectors for gene expression in mammalian cells: Coronavirus\n"
     ]
    }
   ],
   "source": [
    "getData()\n",
    "extract()\n",
    "organize()\n",
    "q = ['coronavirus origin',\n",
    "'coronavirus response to weather changes',\n",
    "'coronavirus immunity']\n",
    "retrieve(q)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b1bdb959",
   "metadata": {},
   "source": [
    "References  \n",
    "https://www.kaggle.com/code/basharzaidat/covid-19-tf-idf-bm25-information-retrieval-system#ORGANIZE"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
